Feature/pdfjs viewer

ewilloughby opened the pull request:
Hey @boveus This PR adds the pdfjs viewer  [pdfjs_viewer_rails gem](https://github.com/senny/pdfjs_viewer-rails)  if a work is a pdf via a pdf viewer method in the sirenia/work_show_presenter (in the same fashion the UV is detected) and also adds formatting/styling for the viewer.  There are 2 new overrides, one for the app/views/hyrax/base/show.html.erb which checks if the work is a pdf if so, renders the work viewer full-width .  The 2nd override is app/views/hyrax/file_sets/media_display/_pdf.html.erb and adds the iframe where pdfjs will render the file.

Lastly - the code at the top of app/views/hyrax/base/show.html.erb is from the University of Albany and passes searches done in hyrax through to pdf.js

boveus commented on app/assets/stylesheets/sirenia/custom.scss:
```diff
@@ -0,0 +1,12 @@
+  iframe#pdfViewer {
+    height: 600px;
+    }
+  }
```
I think this CSS should go into `sirenia/base/show.css` or somewhere else that mimics the location of the views.  That way it will be easier for us to find it if we want to modify it in the future.

Using `height: 50vh;` or something similar instead of static pixels will help to make the pdf viewer more responsive.

ewilloughby replied:
Done.

boveus commented on app/presenters/sirenia/work_show_presenter.rb:
```diff
@@ -5,6 +5,12 @@ class WorkShowPresenter < Hyrax::WorkShowPresenter
+      representative_id.present? &&
+      representative_presenter.present? &&
+      representative_presenter.pdf?
+    end 
```
This method is good.

boveus commented on app/views/hyrax/base/show.html.erb:
```diff
@@ -0,0 +1,69 @@
+<% if breadcrumbs[1] %>
+  <% if breadcrumbs[1].try(:path) %>
+    <% if breadcrumbs[1].path.split("q=")[1] %>
+      <script>
```
I think adding the breadcrumbs value as a hidden attribute on the page with an id of 'query_term' or something and having the JS retrieve that will help us clean the code up a bit.  

You can also move the breadcrumbs logic to the presenter.  I think you could call something like `<% @presenter.query_html(breadcrumbs[1]) %>` here and have all of the logic in the presenter.  Ideally it would be in the controller, but I think it makes sense to throw it into the presenter instead so we don't have to override the controller.  Then it can return something like this: `"<p id='queryterm' hidden> #{query_term} <p>".html_safe`.  

My rationale here is in the event we need to change this page in the future based on any sort of changes made to the core Hyrax code we will be able to do that more easily if it is in our presenter instead of the view we are overriding.

ewilloughby replied:
I moved this code into a partial in the sirenia views called _pdfsearch.html.erb.  This can be nested deeper, probably, but for now it's just in our views folder.

boveus commented on app/views/hyrax/base/show.html.erb:
```diff
@@ -0,0 +1,69 @@
+            $(pdfjsframe.contentWindow.PDFViewerApplication.findBar.findField).val("<%= breadcrumbs[1].path.split("q=")[1].gsub '+', ' ' %>");
+            pdfjsframe.contentWindow.PDFViewerApplication.findBar.dispatchEvent('');
+          });
+        });
```
The JS should go into a javascript file on its own in `app/assets/javascripts/sirenia/pdf_viewer.js` or something like that.  This will add a little complication in how we can retrieve the search parameter from the view, but it will make it easier to manage this code in the future.  Instead of interpolating the ruby into the view, we can instead have the JS retrieve the element's value (see comments above) based on its ID.

ewilloughby replied:
This js has been moved to pdfsearch.js in our javascripts/sirenia folder.  I also made the capturing of the search term(s) pure JS instead of ruby since it was on the page in the breadcrumb, and parsed out the query terms there.

boveus commented on app/views/hyrax/file_sets/media_display/_pdf.html.erb:
```diff
@@ -0,0 +1,27 @@
+                    alt: "",
+                    role: "presentation" %>-->
+
+      <!-- PDF.js iframe -->
```
We should clean up these comments before merging this PR.

ewilloughby replied:
Comments removed.

ewilloughby commented:
I just pushed some commits to address these. I moved the js to a separate file in our javascripts/sirenia folder.  I modified the js to handle all of the capturing of the breadcrumb URL and parsing out the query string instead of it being done in Ruby.  I changed the css to use vh instead of fixed pixel heights, and added a view partial  (views/sirenia/_pdfsearch.htm.erb) that calls the js..  This could be nested deeper in folders to make the organization more clear.  The views/hyrax/base/show.htm.erb now just renders the partial if the work is using a pdf viewer, so it should only run the js to pass along the query terms if the work is a pdf. I'm fine if you need to adjust any of this, but it works.  I'm off now, see you in a couple weeks!
